### PR TITLE
[XLA GPU] Support for mix type gemm bias addition fusion

### PR DESCRIPTION
**Pull request by Cjkkkk**
Considering a gemm  with `D = alpha * A * B + beta * C` form, We observed when `A` and `C` has different type, in XLA HLO, it will appear as 3 HLO instructions: `Add(Convert(Gemm(A, B)), C)`.  

However [cuBLAS](https://docs.nvidia.com/cuda/cublas/index.html#cublasgemmex) and [cuBLASLt](https://docs.nvidia.com/cuda/cublas/index.html#cublasltmatmul) both supports gemm with different types for A and C. To be more specific, both supports 14 different type combinations `(ComputeType, alpha/beta Type, A/BType, C/DType)` .Therefore we can just create a single custom call `Gemm(A, B, C)` to fuse these 3 instructions and call corresponding gemm routines. This fusion  opportunities is observed in weight gradient accumulation scenario. 

This PR is intended to add support for these mix type gemm bias addition fusion, even though cuBLAS and cuBLASLt supports 14 combinations right now. XLA does not have support for choosing different compute type and scale type for different A/C type. We are not planning to add support for choosing different compute type and scale type here because we are not sure how that might affect the precision. Therefore, we plan to add support for 2 cases of form `(ComputeType, alpha/beta Type, A/BType, C/DType)`: 
1. (fp32, fp32, fp16, fp32)
2. (fp32, fp32, bf16, fp32)
3. (fp32, fp32, s8, fp32)

This works for both legacy cuBLAS and cuBLASLt, and it should provide support for both regular gemm and batched gemm.

### CODE REVIEW

File: xla/service/gpu/matmul_utils.cc [hunk 1]
```diff
@@ -620,62 +621,62 @@ Status RunGemm(const GemmConfig& config, se::DeviceMemoryBase lhs_buffer,
...
+        batch_size, m, n, k, lhs, rhs, output,
+        static_cast<int32_t>(config.alpha.real()),
+        static_cast<int32_t>(config.beta), stream, *algorithm,
+        se::blas::kDefaultComputePrecision, profile_result);
   }
+
+  // TYPED_GEMM(F16, F16, F16, F16)
+  TYPED_GEMM(F32, BF16, BF16, BF16)
+  TYPED_GEMM(F32, F16, F16, F16)
+  // TYPED_GEMM(F32, S8, S8, F32)
```
Comment (SandSnip3r) [inline xla/service/gpu/matmul_utils.cc:665]: Why are some things commented out?

**Comment by SandSnip3r** [comment]
@cheshire Isn't this pr a bit moot if Triton already handles all of our mixed type gemms?

**Comment by cheshire** [comment]
@SandSnip3r  It's not, it would be still very useful to get  a roofline reference.

**Comment by cheshire** [comment]
I think this is for different mixed precision case, where A and C have different types. We're mostly interested in cases where A and B have different datatypes, are those supported? The bias fusion isn't super-useful unless it can also fuse the broadcast.

**Comment by Cjkkkk** [comment]
> @cheshire Isn't this pr a bit moot if Triton already handles all of our mixed type gemms?

could you point me to reference of what mix type cases Triton supports now?

**Comment by Cjkkkk** [comment]
> I think this is for different mixed precision case, where A and C have different types. We're mostly interested in cases where A and B have different datatypes, are those supported? The bias fusion isn't super-useful unless it can also fuse the broadcast.

This MR is for cases where A and C have different types. Does Triton support this? This happens a lot in our internal model weight grad accumulation. The weight grad is fp16 and the accumulation is fp32. 

As for the bias that is broadcast, for cublasLT, it can fuse such case. considering a case like this where A [8, 16] fp16, B [16, 32] fp16 and C [32] fp32. We can fuse A@ B + C into one gemm call.

File: xla/service/gpu/matmul_utils.cc (hunk 1 above)
Comment (Cjkkkk) [reply xla/service/gpu/matmul_utils.cc]: removed. Cases are not supported.

**Comment by jprabhas** [comment]
> I think this is for different mixed precision case, where A and C have different types. We're mostly interested in cases where A and B have different datatypes, are those supported? The bias fusion isn't super-useful unless it can also fuse the broadcast.

No, this only supports cases where A(B) has a different type from C. However, C may not only correspond to Bias. The add operation in matmul + C may come from arbitrary add operations. For instance, we found case in weight-gradient accumulation in backward graphs (for T5X and GPT models) where the gradient matmul is in bf16 , while the accumulation of the gradient buffer is done in bf32. This fix fuses those cases and thus voiding separate kernels for add,  giving us 2% perf improvement with T5X and estimated 6% perf improvement with GPT model.

File: xla/service/gpu/gemm_rewriter.cc [hunk 2]
```diff
@@ -1516,7 +1617,7 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     const Shape &output_shape = instr.shape();
 
     TF_ASSIGN_OR_RETURN(bool types_are_supported_by_cublas_lt,
-                        TypesAreSupportedByCublasLt(instr));
+                        AreTypesSupportedByCublasLt(instr));
```
Comment (SandSnip3r) [inline xla/service/gpu/gemm_rewriter.cc:1620]: nit: any reason to change the function name from `TypesAre` to `AreTypes`? I think the original reads better, especially now with the inconsistent naming of the bool.

**Review by SandSnip3r — APPROVED** [review]

File: xla/service/gpu/gemm_rewriter.cc (hunk 2 above)
Comment (Cjkkkk) [reply xla/service/gpu/gemm_rewriter.cc]: Not really. Just renamed to `AreTypes*`

**Review by akuegel — APPROVED** [review]

**Review by SandSnip3r — APPROVED** [review]

**Review by SandSnip3r — APPROVED** [review]

**Comment by akuegel** [comment]
> the internal XLA Linux CPU and GPU failed: `ERROR: /root/.cache/bazel/_bazel_root/217377b0e928b171b843eb11ea7bc36e/external/llvm-project/llvm/BUILD.bazel:200:11: no such package '@llvm_zlib//': The repository '@llvm_zlib' could not be resolved: Repository '@llvm_zlib' is not defined and referenced by '@llvm-project//llvm:Support'` How do we resolve this? Did not encounter this before.

This was a temporary breakage of all Kokoro jobs, it should be resolved already.

**Comment by Cjkkkk** [comment]
> Our internal integration tests show some errors:
> 
> 
> Looks like sometimes an f16 operand is aliased with an f32 output after this change?
> 
> RET_CHECK failure ([third_party/tensorflow/compiler/xla/service/hlo_verifier.cc:1330](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/service/hlo_verifier.cc?l=1330&ws=tap-presubmit-rerun-server/4095624&snapshot=2)) operand_subshape == output_subshape Different aliasing shapes: f16[32,128] vs f32[32,128]
> 
> 
> Not really sure about this: XlaRuntimeError: INTERNAL: AsBlasDataType: unsupported type [UNKNOWN]
> 
> 
> In some tests the precision decreased: "0.011279851 not less than or equal to 0.006105898" "13.649316 not less than or equal to 0.05"
> 
> I can't easily give you full information, because most of these tests are in a silo that I don't have permission to read, but perhaps the first one is specific enough to look into it.

I pushed a new commit which should fix the first issue. For the second issue, I didn't change any logic in AsBlasDataType but I enhenced the error message in AsBlasDataType to show which primitiveType caused the fail. For the last issue, it would be helpful to have the actual test cases.

**Comment by Cjkkkk** [comment]
Having another issue in Linux build: 
`Error in patch: Error applying patch /root/.cache/bazel/_bazel_root/217377b0e928b171b843eb11ea7bc36e/external/tsl/third_party/llvm/zstd.patch: Expecting more chunk line at line 19
ERROR: no such package '@llvm-raw//utils/bazel': Error applying patch /root/.cache/bazel/_bazel_root/217377b0e928b171b843eb11ea7bc36e/external/tsl/third_party/llvm/zstd.patch: Expecting more chunk line at line 19`

**Comment by SandSnip3r** [comment]
```
HloModule test

ENTRY main {
  param_0 = f32[240,88]{1,0} parameter(0)
  param_1 = f32[88,4]{1,0} parameter(1)
  dot = f32[240,4]{1,0} dot(param_0, param_1), lhs_contracting_dims={1}, rhs_contracting_dims={0}, operand_precision={highest,highest}
  constant_255 = f32[] constant(255)
  broadcast = f32[240,4]{1,0} broadcast(constant_255), dimensions={}
  multiply = f32[240,4]{1,0} multiply(dot, broadcast)
  ROOT result = u8[240,4]{1,0} convert(multiply)
}
```

**Comment by Cjkkkk** [comment]
> Getting an error with this:
> 
> ```
> HloModule test
> 
> ENTRY main {
>   param_0 = f32[240,88]{1,0} parameter(0)
>   param_1 = f32[88,4]{1,0} parameter(1)
>   dot = f32[240,4]{1,0} dot(param_0, param_1), lhs_contracting_dims={1}, rhs_contracting_dims={0}, operand_precision={highest,highest}
>   constant_255 = f32[] constant(255)
>   broadcast = f32[240,4]{1,0} broadcast(constant_255), dimensions={}
>   multiply = f32[240,4]{1,0} multiply(dot, broadcast)
>   ROOT result = u8[240,4]{1,0} convert(multiply)
> }
> ```
> 
> EDIT: This is the root of @tdanyluk's #2 in the comment above.

issue 1 and 2 should be resolved. could you help confirm this?

**Comment by jprabhas** [comment]
> > Getting an error with this:
> > ```
> > HloModule test
> > 
> > ENTRY main {
> >   param_0 = f32[240,88]{1,0} parameter(0)
> >   param_1 = f32[88,4]{1,0} parameter(1)
> >   dot = f32[240,4]{1,0} dot(param_0, param_1), lhs_contracting_dims={1}, rhs_contracting_dims={0}, operand_precision={highest,highest}
> >   constant_255 = f32[] constant(255)
> >   broadcast = f32[240,4]{1,0} broadcast(constant_255), dimensions={}
> >   multiply = f32[240,4]{1,0} multiply(dot, broadcast)
> >   ROOT result = u8[240,4]{1,0} convert(multiply)
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > EDIT: This is the root of @tdanyluk's #2 in the comment above.
> 
> issue 1 and 2 should be resolved. could you help confirm this?

It would be good to add test cases for these issues in the OSS . These will be more visible than the internal tests.

**Comment by Cjkkkk** [comment]
could you provide the type combination for the precision loss cases?

**Comment by Cjkkkk** [comment]
> > It would be good to add test cases for these discovered issues in the OSS as that will be more visible than the internal tests.
> 
> This PR seems like a good starting point. This type of thing ought to be covered in gemm_rewrite_test

I add tests to cover the issues google internal tests revealed. It seems there are still internal tests failing, could you help show those cases.

**Comment by cheshire** [comment]
I see this failure in OSS build (no idea why you don't see it directly):

```
[ RUN      ] TritonAutotunerTest.Int8FusedGemm256

// CHECK: ROOT %out.1 = f16[128,6144]{1,0} dot(%c.1, %parameter_1), lhs_contracting_dims={1}, rhs_contracting_dims={0}
          ^
<stdin>:3:31: note: scanning from here
%triton_gemm_out_computation (parameter_0: s8[128,256], parameter_1: f16[256,6144]) -> f16[4,128,6144] {
                              ^
<stdin>:9:2: note: possible intended match here
 ROOT %dot = f16[4,128,6144]{2,1,0} dot(%bitcast, %bitcast.1), lhs_batch_dims={1}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={1}
 ^

Input file: <stdin>
Check file: /b/f/w/_tmp/9113fa3f25a5251c4bf73102457874b9/tempfile-localhost-a94aa212-27-5fc725ad05953

-dump-input=help explains the following input dump.

Input was:
<<<<<<
           1: HloModule module, entry_computation_layout={(s8[128,256]{1,0}, f16[256,6144]{1,0})->f16[128,6144]{1,0}} 
           2:  
           3: %triton_gemm_out_computation (parameter_0: s8[128,256], parameter_1: f16[256,6144]) -> f16[4,128,6144] { 
check:2       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:3'0                                   X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           4:  %parameter_0 = s8[128,256]{1,0} parameter(0) 
check:3'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           5:  %c.1 = f16[128,256]{1,0} convert(%parameter_0) 
check:3'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           6:  %bitcast = f16[128,4,64]{2,1,0} bitcast(%c.1) 
check:3'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           7:  %parameter_1 = f16[256,6144]{1,0} parameter(1) 
check:3'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           8:  %bitcast.1 = f16[4,64,6144]{2,1,0} bitcast(%parameter_1) 
check:3'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           9:  ROOT %dot = f16[4,128,6144]{2,1,0} dot(%bitcast, %bitcast.1), lhs_batch_dims={1}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={1}
```

**Comment by Cjkkkk** [comment]
> b

unable to produce this on my machine, this test passed successfully. However, `TritonAutotunerTest.SelectsSplitK ` seems to have some precision loss. Were you able to see this?

**Comment by Cjkkkk** [comment]
> I see this failure in OSS build (no idea why you don't see it directly):
> 
> ```
> [ RUN      ] TritonAutotunerTest.Int8FusedGemm256
> 
> // CHECK: ROOT %out.1 = f16[128,6144]{1,0} dot(%c.1, %parameter_1), lhs_contracting_dims={1}, rhs_contracting_dims={0}
>           ^
> <stdin>:3:31: note: scanning from here
> %triton_gemm_out_computation (parameter_0: s8[128,256], parameter_1: f16[256,6144]) -> f16[4,128,6144] {
>                               ^
> <stdin>:9:2: note: possible intended match here
>  ROOT %dot = f16[4,128,6144]{2,1,0} dot(%bitcast, %bitcast.1), lhs_batch_dims={1}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={1}
>  ^
> 
> Input file: <stdin>
> Check file: /b/f/w/_tmp/9113fa3f25a5251c4bf73102457874b9/tempfile-localhost-a94aa212-27-5fc725ad05953
> 
> -dump-input=help explains the following input dump.
> 
> Input was:
> <<<<<<
>            1: HloModule module, entry_computation_layout={(s8[128,256]{1,0}, f16[256,6144]{1,0})->f16[128,6144]{1,0}} 
>            2:  
>            3: %triton_gemm_out_computation (parameter_0: s8[128,256], parameter_1: f16[256,6144]) -> f16[4,128,6144] { 
> check:2       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> check:3'0                                   X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
>            4:  %parameter_0 = s8[128,256]{1,0} parameter(0) 
> check:3'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            5:  %c.1 = f16[128,256]{1,0} convert(%parameter_0) 
> check:3'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            6:  %bitcast = f16[128,4,64]{2,1,0} bitcast(%c.1) 
> check:3'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            7:  %parameter_1 = f16[256,6144]{1,0} parameter(1) 
> check:3'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            8:  %bitcast.1 = f16[4,64,6144]{2,1,0} bitcast(%parameter_1) 
> check:3'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>            9:  ROOT %dot = f16[4,128,6144]{2,1,0} dot(%bitcast, %bitcast.1), lhs_batch_dims={1}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={1}
> ```

unable to produce this on my machine, this test passed successfully. However, TritonAutotunerTest.SelectsSplitK  seems to have some precision loss. Were you able to see this?

**Comment by Cjkkkk** [comment]
> feedback/copybara — Google internal checks FAILED for runs with create time 2023-05-25T19:24:46.23

i am not able to see the details. Could you share the current status of the internal test.

**Comment by SandSnip3r** [comment]
> > feedback/copybara — Google internal checks FAILED for runs with create time 2023-05-25T19:24:46.23
> 
> i am not able to see the details. Could you share the current status of the internal test.

I see gemm_rewriter_tests is failing on LegacyCublasGemmRewriteTest.MatrixBiasMixTypeBatched. There's a numeric issue.


```
[third_party/tensorflow/compiler/xla/service/gpu/tests/gemm_rewrite_test.cc:1816](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/xla/service/gpu/tests/gemm_rewrite_test.cc?l=1816&ws=tap-presubmit-server/53720372&snapshot=2): Failure
Value of: RunAndCompare(hlo_text, ErrorSpec{1e-3, 1e-3})
  Actual: false (
Mismatch count 1020 (99.6094%) in shape f32[4,16,16] (1024 elements), abs bound 0.001, rel bound 0.001
Top relative error mismatches:
  actual    0.0716888085, expected  8.85352492e-05, index {1,9,0}, rel error      809, abs error   0.0716
  actual   -0.0967227593, expected  0.000445209444, index {2,2,10}, rel error      218, abs error   0.0972
  actual    -0.147521079, expected  -0.00103670359, index {2,13,14}, rel error      141, abs error    0.146
  actual    -0.146679044, expected  -0.00117123127, index {2,0,3}, rel error      124, abs error    0.146
  actual     0.151325643, expected  -0.00134796649, index {0,12,15}, rel error      113, abs error    0.153
Absolute magnitude breakdown of actual values:
  0      <= x < 0.0001 :       0 (  0.0000%)
  0.0001 <= x < 0.001  :       8 (  0.7812%), mismatches 8
  0.001  <= x < 0.01   :      60 (  5.8594%), mismatches 60
  0.01   <= x < 0.1    :     586 ( 57.2266%), mismatches 582
  0.1    <= x < 1      :     370 ( 36.1328%), mismatches 370
  1      <= x < inf    :       0 (  0.0000%)
Elements exceeding abs error bound 0.001: 1020 (99.6094%)
Relative error breakdown of elements exceeding abs error bound:
  <  0.0001 :       0 (0.0000%)
  >= 0.0001 :    1020 (100.0000%)
  >= 0.001  :    1020 (100.0000%)
  >= 0.01   :    1020 (100.0000%)
  >= 0.1    :     962 (94.3137%)
  >= 1      :     455 (44.6078%)
Elements exceeding rel error bound 0.001: 1024 (100.0000%)
Absolute error breakdown of elements exceeding rel error bound:
  <  0.0001 :       1 (0.0977%)
  >= 0.0001 :    1023 (99.9023%)
  >= 0.001  :    1020 (99.6094%)
  >= 0.01   :     938 (91.6016%)
  >= 0.1    :     332 (32.4219%)
  >= 1      :       0 (0.0000%)


Expected literal:
[TRUNCATED, Literal with more than 1000 values]

Actual literal:
[TRUNCATED, Literal with more than 1000 values])
Expected: true
```

Can you repro?

**Comment by Cjkkkk** [comment]
> > > feedback/copybara — Google internal checks FAILED for runs with create time 2023-05-25T19:24:46.23
> > 
> > 
> > i am not able to see the details. Could you share the current status of the internal test.
> 
> I see gemm_rewriter_tests is failing on LegacyCublasGemmRewriteTest.MatrixBiasMixTypeBatched. There's a numeric issue.
> 
> ```
> [xla/service/gpu/tests/gemm_rewrite_test.cc:1816]: Failure
> Value of: RunAndCompare(hlo_text, ErrorSpec{1e-3, 1e-3})
>   Actual: false (
> Mismatch count 1020 (99.6094%) in shape f32[4,16,16] (1024 elements), abs bound 0.001, rel bound 0.001
> Top relative error mismatches:
>   actual    0.0716888085, expected  8.85352492e-05, index {1,9,0}, rel error      809, abs error   0.0716
>   actual   -0.0967227593, expected  0.000445209444, index {2,2,10}, rel error      218, abs error   0.0972
>   actual    -0.147521079, expected  -0.00103670359, index {2,13,14}, rel error      141, abs error    0.146
>   actual    -0.146679044, expected  -0.00117123127, index {2,0,3}, rel error      124, abs error    0.146
>   actual     0.151325643, expected  -0.00134796649, index {0,12,15}, rel error      113, abs error    0.153
> Absolute magnitude breakdown of actual values:
>   0      <= x < 0.0001 :       0 (  0.0000%)
>   0.0001 <= x < 0.001  :       8 (  0.7812%), mismatches 8
>   0.001  <= x < 0.01   :      60 (  5.8594%), mismatches 60
>   0.01   <= x < 0.1    :     586 ( 57.2266%), mismatches 582
>   0.1    <= x < 1      :     370 ( 36.1328%), mismatches 370
>   1      <= x < inf    :       0 (  0.0000%)
> Elements exceeding abs error bound 0.001: 1020 (99.6094%)
> Relative error breakdown of elements exceeding abs error bound:
>   <  0.0001 :       0 (0.0000%)
>   >= 0.0001 :    1020 (100.0000%)
>   >= 0.001  :    1020 (100.0000%)
>   >= 0.01   :    1020 (100.0000%)
>   >= 0.1    :     962 (94.3137%)
>   >= 1      :     455 (44.6078%)
> Elements exceeding rel error bound 0.001: 1024 (100.0000%)
> Absolute error breakdown of elements exceeding rel error bound:
>   <  0.0001 :       1 (0.0977%)
>   >= 0.0001 :    1023 (99.9023%)
>   >= 0.001  :    1020 (99.6094%)
>   >= 0.01   :     938 (91.6016%)
>   >= 0.1    :     332 (32.4219%)
>   >= 1      :       0 (0.0000%)
> 
> 
> Expected literal:
> [TRUNCATED, Literal with more than 1000 values]
> 
> Actual literal:
> [TRUNCATED, Literal with more than 1000 values])
> Expected: true
> ```
> 
> Can you repro?

```
[       OK ] Fp8CublasTestsBothLegacyAndLt/ParameterizedFp8GemmRewriteTest.ScaledABUnscaledDF8TF32E5M2/1 (9 ms)                           │·················································
[----------] 52 tests from Fp8CublasTestsBothLegacyAndLt/ParameterizedFp8GemmRewriteTest (1471 ms total)                                  │·················································
                                                                                                                                          │·················································
[----------] Global test environment tear-down                                                                                            │·················································
[==========] 188 tests from 6 test suites ran. (17307 ms total)                                                                           │·················································
[  PASSED  ] 187 tests.                                                                                                                   │·················································
[  SKIPPED ] 1 test, listed below:                                                                                                        │·················································
[  SKIPPED ] GemmRewriteTest.TestBatchedAutotuning 
```
Not able to repro this, what is the gpu and cublas version? I am using compute capability 8.7 and cublas 12.1.

**Comment by cheshire** [comment]
Internally, I see numeric failure on `LegacyCublasGemmRewriteTest.MatrixBiasMixTypeBatched`, with an absolute difference around ~0.2 .

**Comment by cheshire** [comment]
(we're running CUDA11.8)

**Comment by cheshire** [comment]
I see another miscompare on Ampere.

**Comment by Cjkkkk** [comment]
> Hi @Cjkkkk, we are still seeing test failures on LegacyCublasGemmRewriteTest.MatrixBiasMixTypeBatched still. Can you confirm this test passes with the version information provided above?

Hi, I tried cuda 11.8 and Ampere arch. I still can not reproduce this on my local machine. Test passed successfully.

**Comment by cheshire** [comment]
> Hi, I tried cuda 11.8 and Ampere arch. I still can not reproduce this on my local machine. Test passed successfully.

Maybe Pascal/Volta?

**Comment by Cjkkkk** [comment]
> > Hi, I tried cuda 11.8 and Ampere arch. I still can not reproduce this on my local machine. Test passed successfully.
> 
> Maybe Pascal/Volta?

Have some issues setting up pascal/volta machine, will get back to you once I have the results

**Comment by Cjkkkk** [comment]
> MatrixBiasMixTypeBatched

Passed on Turing/Volta arch, still testing on pascal

File: xla/service/gpu/matmul_utils.cc (hunk 1 above)
Comment (akuegel) [reply xla/service/gpu/matmul_utils.cc]: It seems you removed it here, but not in the TypesAreSupportedByLegacyCublas function. In some internal tests, we see this error being triggered:
"Unexpected GEMM dtype: s8 s8 f32".
Can you please double-check all supported cases again, whether they have a corresponding TYPED_GEMM or TYPED_GEMM_COMPLEX macro call?

File: xla/service/gpu/tests/gemm_rewrite_test.cc [hunk 3]
```diff
@@ -1184,6 +1275,7 @@ class LegacyCublasGemmRewriteTest : public GemmRewriteTest {
   DebugOptions GetDebugOptionsForTest() override {
     DebugOptions debug_options = GemmRewriteTest::GetDebugOptionsForTest();
     debug_options.set_xla_gpu_enable_cublaslt(false);
+    debug_options.set_xla_gpu_simplify_all_fp_conversions(true);
```
Comment (SandSnip3r) [inline xla/service/gpu/tests/gemm_rewrite_test.cc:1278]: Why enable this for all tests?

Comment (Cjkkkk) [reply xla/service/gpu/tests/gemm_rewrite_test.cc:1278]: cuz mix type gemm test is guarded by this flag, this is not enabled in google internal test somehow. And it seems it is enabled by default when i ran it on my local machine

Comment (Cjkkkk) [reply xla/service/gpu/tests/gemm_rewrite_test.cc]: removed. Turns out it is related to filecheck.

File: xla/service/gpu/matmul_utils.cc (hunk 1 above)
Comment (Cjkkkk) [reply xla/service/gpu/matmul_utils.cc]: it is added back.

**Pull request closed without merging**